docs/development/defaulting.md

```diff
@@ -0,0 +1,29 @@
+# Defaulting Strategy and Developer Guideline
+
+This document walks you through:
+
+- Conventions to be followed when writing defaulting functions
+- How to write test for defaulting function
+
+The document is aimed towards developers that want to contribute code and need to write defaulting code and unit tests covering the defaulting functions, as well as maintainers and reviewers that review code.
+It serves as a common guide that we commit to follow in our project to ensure consistency in our defaulting code, good coverage for high confidence, and good maintainability.
+
+## Writing defaulting code
+
+- Every kubernetes type should have a dedicated `defaults_*.go` file. For instance, if you have a [`Shoot`](../../pkg/apis/core/types_shoot.go) type, there should be a corresponding [`defaults_shoot.go`](../../pkg/apis/core/v1beta1/defaults_shoot.go) file containing all defaulting logic for that type.
+- If there is only one type under an api group then we can just have `types.go` and a corresponding `defaults.go`. For instance, `resourcemanager` api has only one [`types.go`]((../../pkg/resourcemanager/apis/config/v1alpha1/types.go)), hence in this case only [`defaults.go`](../../pkg/resourcemanager/apis/config/v1alpha1/defaults.go) file would suffice. 
+- Aim to segregate each struct type into its own `SetDefaults_*` function. These functions encapsulate the defaulting logic specific to the corresponding struct type, enhancing modularity and maintainability. For example, [`ServerConfiguration`](https://github.com/gardener/gardener/blob/ff5a5be6049777b0695659a50189e461e1b17796/pkg/resourcemanager/apis/config/v1alpha1/types.go#L64-L74) struct in `resourcemanager` api has corresponding [`SetDefaults_ServerConfiguration()`](https://github.com/gardener/gardener/blob/ff5a5be6049777b0695659a50189e461e1b17796/pkg/resourcemanager/apis/config/v1alpha1/defaults.go#L73-L92) function.
+
+:warning: Ensure to run the `make generate WHAT=codegen` command whenever any changes are done in defaulting logic or a new `SetDefaults_*` function is added, which generates the `zz_generated.defaults.go` file containing the overall defaulting function.
```

Comment:
I am not sure about ` whenever any changes are done in defaulting logic`? What do you mean with it? It is enough to run the generation only when a `SetDefaults_*` function is added or removed?